crates/noirc_frontend/src/parser/errors.rs

```diff
@@ -1,19 +1,41 @@
 use crate::lexer::token::Token;
-use crate::BinaryOp;
-use late_alloc_set::LateAllocSet;
+use crate::Expression;
+use small_ord_set::SmallOrdSet;
+use thiserror::Error;
 
 use iter_extended::vecmap;
 use noirc_errors::CustomDiagnostic as Diagnostic;
 use noirc_errors::Span;
 
-use super::labels::ParserLabel;
-
-mod late_alloc_set;
+use super::labels::ParsingRuleLabel;
+
+#[derive(Debug, Clone, PartialEq, Eq, Error)]
+pub enum ParserErrorReason {
+    #[error("Arrays must have at least one element")]
+    ZeroSizedArray,
+    #[error("Unexpected '{0}', expected a field name")]
+    ExpectedFieldName(Token),
+    #[error("Expected a ; separating these two statements")]
+    MissingSeparatingSemi,
+    #[error("constrain keyword is deprecated")]
+    ConstrainDeprecated,
+    #[error("Expression is invalid in an array-length type: '{0}'. Only unsigned integer constants, globals, generics, +, -, *, /, and % may be used in this context.")]
+    InvalidArrayLengthExpression(Expression),
+}
 
+/// Represents a parsing error, or a parsing error in the making.
+///
+/// `ParserError` is used extensively by the parser, as it not only used to report badly formed
+/// token streams, but also as a general intermediate that accumulates information as various
+/// parsing rules are tried. This struct is constructed and destructed with a very high frequency
+/// and as such, the time taken to do so significantly impacts parsing performance. For this
+/// reason we use `SmallOrdSet` to avoid heap allocations for as long as possible - this greatly
+/// inflates the size of the error, but this is justified by a resulting increase in parsing
+/// speeds of approximately 40% in release mode.
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub struct ParserError {
-    expected_tokens: LateAllocSet<Token>,
-    expected_labels: LateAllocSet<ParserLabel>,
+    expected_tokens: SmallOrdSet<[Token; 3]>,
+    expected_labels: SmallOrdSet<[ParsingRuleLabel; 3]>,
```

Comment:
Originally it was because I found that 90% of errors go accumulate 3 tokens. However, I've just sampled at different sizes and found that they might as well be sized 1.